arbnode/batch_poster.go

```diff
@@ -563,29 +563,29 @@ var EspressoFetchTransactionErr = errors.New("failed to fetch the espresso trans
 
 // Adds a block merkle proof to an Espresso justification, providing a proof that a set of transactions
 // hashes to some light client state root.
-func (b *BatchPoster) checkEspressoValidation() error {
+func (b *BatchPoster) checkEspressoValidation() bool {
+	b.building.segments.SetWaitingForValidation()
 	if b.streamer.espressoClient == nil && b.streamer.lightClientReader == nil {
-		// We are not using espresso mode since these haven't been set
-		return nil
+		// We are not using espresso mode since these haven't been set, return true to advance batch posting
+		return true
+	}
+	if b.streamer.EscapeHatchEnabled {
+		log.Warn("skipped espresso verification due to hotshot failure", "pos", b.building.msgCount)
+		return true // return true to skip verification of batch
 	}
-
 	lastConfirmed, err := b.streamer.getLastConfirmedPos()
 	if err != nil {
 		log.Error("failed call to get last confirmed pos", "err", err)
-		return err
+		return false // if we get an error we can't validate
 	}
 
 	// This message has passed the espresso verification
-	if lastConfirmed != nil && b.building.msgCount <= *lastConfirmed {
-		return nil
-	}
 
-	if b.streamer.EscapeHatchEnabled {
-		log.Warn("skipped espresso verification due to hotshot failure", "pos", b.building.msgCount)
-		return nil
+	if lastConfirmed != nil && b.building.msgCount-1 <= *lastConfirmed {
```

Comment:
I am a bit worried about this change. With this implementation the last confirmed was always ending one behind `b.building.msgCount` I suspect that this is just an off by one error because the 3.3.2 code also doesn't handle including the message at msgCount in the batch. It also makes me wonder why this check worked in the first place.

I do think we should add some tests around different transaction loads with different config settings for the batch poster so we can at least sniff out some of these weird edge cases with how we build the batch.

Reply:
It worked because `checkEspressoValidation` was running in a loop where `b.building.msgCount < msgCount` was holding true with every iteration and `		b.building.msgCount++` was executing at the end of each iteration. So we were essentially checking that with each iteration because the b.building.msgCount is increasing `b.building.msgCount-1 <= *lastConfirmed` last confirmed is always greater or equal. Let me know if that makes sense!

Reply:
So it's just because we switched the order between checking espresso validation and incrementing the counter, and so the old check is equivalent to the new one? That makes sense

Reply:
In any case the new one is much more intuitive I think: `b.building.msgCount-1` should be the *position* of the last message in the batch and we are simply checking that this message position is confirmed